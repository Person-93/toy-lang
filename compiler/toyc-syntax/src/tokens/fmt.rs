use crate::tokens::{NumLitPrefix, NumLitType};
use std::fmt::{Debug, Display, Formatter};

impl Display for super::Ident {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", self.0)
  }
}

impl Debug for super::Ident {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "Ident({self})")
  }
}

impl Display for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    Display::fmt(&self.0, f)
  }
}

impl Debug for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "StrLit({self})")
  }
}

impl Display for super::NumLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    let super::NumLit {
      prefix,
      val,
      decimal,
      ty,
    } = self;

    let prefix = match prefix {
      None => "",
      Some(NumLitPrefix::Hex) => "0x",
      Some(NumLitPrefix::Binary) => "0b",
    };

    let decimal = match decimal {
      None => String::new(),
      Some(n) => format!("{n}"),
    };

    let ty = match ty {
      None => "",
      Some(NumLitType::Float) => "f",
      Some(NumLitType::Int) => "i",
      Some(NumLitType::Unsigned) => "u",
    };

    write!(f, "{prefix}{val}{decimal}{ty}")
  }
}
