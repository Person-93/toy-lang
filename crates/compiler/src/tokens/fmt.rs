use crate::tokens::{NumLitPrefix, NumLitType};
use std::fmt::{Display, Formatter};

impl Display for super::StrLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "\"{}\"", self.0)
  }
}

impl Display for super::BoolLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", if self.0 { "true" } else { "false" })
  }
}

impl Display for super::NumLit {
  fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
    let super::NumLit {
      prefix,
      val,
      decimal,
      ty,
    } = self;

    let prefix = match prefix {
      None => "",
      Some(NumLitPrefix::Hex) => "0x",
      Some(NumLitPrefix::Binary) => "0b",
    };

    let decimal = match decimal {
      None => String::new(),
      Some(n) => format!("{n}"),
    };

    let ty = match ty {
      None => "",
      Some(NumLitType::Float) => "f",
      Some(NumLitType::Int) => "i",
      Some(NumLitType::Unsigned) => "u",
    };

    write!(f, "{prefix}{val}{decimal}{ty}")
  }
}
