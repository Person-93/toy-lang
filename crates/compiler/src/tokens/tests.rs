use super::*;
use std::fmt::{Debug, Formatter};

use insta::{assert_snapshot, glob};
use logos::{Logos, Span};
use std::fs;

#[test]
fn snapshot() {
  glob!("toys/*.toy", |file| {
    let text = fs::read_to_string(file).unwrap();
    assert_snapshot!(Token::lexer(&text)
      .spanned()
      .map(|(token, span)| Output(token, span.clone(), &text[span]).to_string())
      .collect::<Vec<_>>()
      .join("\n"))
  });

  struct Output<'a>(Token, Span, &'a str);

  impl std::fmt::Display for Output<'_> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
      if let Token::Error = self.0 {
        write!(f, "ERROR @ {:?}: \"{}\"", self.1, self.2)
      } else {
        write!(f, "{:?} @ {:?}", self.0, self.1)
      }
    }
  }
}
