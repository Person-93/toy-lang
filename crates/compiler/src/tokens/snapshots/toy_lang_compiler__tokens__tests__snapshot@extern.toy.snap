---
source: crates/compiler/src/tokens/tests.rs
assertion_line: 12
expression: "Token::lexer(&text).spanned().map(|(token, span)|\n                                      Output(token, span.clone(),\n                                             &text[span]).to_string()).collect::<Vec<_>>().join(\"\\n\")"
input_file: toys/extern.toy

---
KwExtern @ 0..6
KwFn @ 7..9
Ident @ 10..16: answer
LeftParen @ 16..17
RightParen @ 17..18
Returns @ 19..21
Ident @ 22..24: i8
Colon @ 24..25
LeftBrace @ 26..27
NumLit @ 28..30: 42
RightBrace @ 31..32
KwExtern @ 34..40
KwFn @ 41..43
Ident @ 44..48: ping
LeftParen @ 48..49
RightParen @ 49..50
Returns @ 51..53
Ident @ 54..59: usize
Colon @ 59..60
LeftBrace @ 61..62
Ident @ 63..67: ping
LeftParen @ 67..68
RightParen @ 68..69
RightBrace @ 70..71
KwExtern @ 73..79
KwFn @ 80..82
Ident @ 83..87: pong
LeftParen @ 87..88
RightParen @ 88..89
Returns @ 90..92
Ident @ 93..98: usize
Semi @ 98..99
